src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TransportCreationHandler.cs

```diff
@@ -0,0 +1,192 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Microsoft.Data.SqlClient.SNI;
+using Microsoft.Data.SqlClientX.Handlers.Connection;
+
+namespace Microsoft.Data.SqlClientX.Handlers.TransportCreation
+{
+    internal class TransportCreationHandler : IHandler<ConnectionHandlerContext>
+    {
+        private const int KeepAliveIntervalSeconds = 1;
+        private const int KeepAliveTimeSeconds = 30;
+
+        #if NET8_0_OR_GREATER
+        private static readonly TimeSpan DefaultPollTimeout = TimeSpan.FromSeconds(30);
+        #else
+        private const int DefaultPollTimeout = 30 * 100000; // 30 seconds as microseconds
+        #endif
+
+        /// <inheritdoc />
+        public IHandler<ConnectionHandlerContext> NextHandler { get; set; }
+
+        /// <inheritdoc />
+        public async ValueTask Handle(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            if (context.DataSource is null)
+            {
+                context.Error = new ArgumentNullException(nameof(context));
+                return;
+            }
+
+            try
+            {
+                // @TODO: Build CoR for handling the different protocols in order
+                if (context.DataSource.ResolvedProtocol is DataSource.Protocol.TCP)
+                {
+                    context.ConnectionStream = await HandleTcpRequest(context, isAsync, ct).ConfigureAwait(false);
+                }
+            }
+            catch (Exception e)
+            {
+                context.Error = e;
+            }
+
+            if (NextHandler is not null)
+            {
+                await NextHandler.Handle(context, isAsync, ct).ConfigureAwait(false);
+            }
+        }
+
+        private ValueTask<Stream> HandleNamedPipeRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private ValueTask<Stream> HandleSharedMemoryRequest()
+        {
+            throw new NotImplementedException();
+        }
+
+        private async ValueTask<Stream> HandleTcpRequest(ConnectionHandlerContext context, bool isAsync, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            // DNS lookup
+            var ipAddresses = isAsync
+                ? await Dns.GetHostAddressesAsync(context.DataSource.ServerName, ct).ConfigureAwait(false)
+                : Dns.GetHostAddresses(context.DataSource.ServerName);
+            if (ipAddresses.Length == 0)
```

Comment:
I think this is something that we will revisit in a future PR. It's definitely something that's churning in the back of my head how we'll handle cancelling if the connection timeout expires. Especially how we may or may not be able to halt synchronous operations.

Out of curiosity, what does NB mean in this context? I'm more familiar with it referring to "no bump"

Reply:
One point I'd been thinking about for the main project (and which might be useful for the future PR) was having a `TimedCancellationToken` readonly struct which combined `CancellationToken` and the time taken before it times out. Having `CancellationToken` and `RemainingTime` properties would be useful for combining the timeout handling. It's a little awkward to retrofit, but probably easier to start working with from the start.

NB is just a side note - *[nota bene](https://en.wikipedia.org/wiki/Nota_bene)*.

Reply:
I was definitely thinking of something similar - if we can combine both the time out and user cancelled exception together it would very much simplify the code. We probably want to be able to throw different exceptions for which token was cancelled, so some more research might be necessary, but I bet someone has done it before

Reply:
This is a *very* rough draft, but my first thought is to have a method closely tied to the connection string which either returns `null` (for a Connection Timeout of zero) or a new instance of `CancellationTokenSource` which has had `CancelAfter(ConnectionTimeout)` called on it. The `SqlConnection.Connect` method could call this to get a CTS, then use `CancellationTokenSource.CreateLinkedTokenSource` to combine this CTS' `CancellationToken` with the user-provided one. The output of this could be threaded through all of the SqlClientX internals.

There's space for some optimisation there: if the ConnectionTimeout-based CancellationTokenSource is null, there's no need to allocate or combine CTSes; all of the CancellationTokenSources need to be disposed of. There's also the need you mentioned to figure out which CancellationToken caused the cancellation exception - I think we'd just check if the user-provided CT's `IsCancellationRequested` property is set. Finally, we might want to track how long the final operation took for telemetry purposes.

Of course, all of this logic applies just as neatly to command timeouts. Perhaps a readonly struct with a better name than `OperationCancellation` - `CancellationOrTimeoutNotifier` is too awkward!